subsys/fs/nvs/nvs.c

```diff
@@ -654,45 +660,104 @@ int nvs_reinit(struct nvs_fs *fs)
 		goto err;
 	}
 
+	if (sectors_in_use == fs->sector_count) {
+		/* one sector should always be empty, unless power was
+		 * cut during garbage collection, clear the last sector and
+		 * start gc again on the entry sector.
+		 */
+		SYS_LOG_DBG("Erasing last sector");
+		addr = fs->entry_sector * fs->sector_size;
+		/* calculate last sector start */
+		_nvs_addr_decrease(fs, &addr, fs->sector_size);
+		/* erase the last sector */
+		rc = _nvs_flash_erase(fs, addr, fs->sector_size);
+		if (rc) {
+			goto err;
+		}
+		/* initialize the erased sector */
+		rc = _nvs_sector_init(fs, addr);
+		if (rc) {
+			goto err;
+		}
+		SYS_LOG_DBG("Restarting garbage collection");
+		addr = fs->entry_sector * fs->sector_size;
+		_nvs_entry_sector_advance(fs);
+		rc = _nvs_gc(fs);
+		if (rc) {
+			goto err;
+		}
+		rc = _nvs_flash_erase(fs, addr, fs->sector_size);
+		if (rc) {
+			goto err;
+		}
+		rc = 0;
+		goto err;
+	}
+
 	/* Valid sectors found in flash, need to set fs->write_location */
 	addr = fs->entry_sector * fs->sector_size;
 	while (1) {
+		off_t entry_end_addr;
+
 		rc = _nvs_flash_rd_len_id_start(fs, addr, &id, &len1);
 		if (rc) {
 			goto err;
 		}
 		if (id == NVS_ID_EMPTY) {
-			break;
+			rc = 0;
+			goto err;
 		}
 		entry_len = get_wbs_aligned_entry_len(fs, len1);
+		/* avoid doing read over sector boundary */
+		entry_end_addr = (fs->write_location & (fs->sector_size - 1))
+				 + entry_len;
+		if (entry_end_addr > fs->sector_size) {
+			break;
+		}
 		_nvs_addr_advance(fs, &addr, entry_len);
 		rc = _nvs_flash_rd_len_id_end(fs, addr, &id, &len2);
 		if (len1 != len2) {
-			rc = -EROFS;
-			goto err;
+			break;
 		}
 		fs->write_location = addr;
 	}
 
-	if (sectors_in_use == fs->sector_count) {
-		/* one sector should always be empty, unless power was
-		 * cut during garbage collection start gc again on the
-		 * entry sector.
-		 */
-		SYS_LOG_DBG("Restarting garbage collection");
-		addr = fs->entry_sector * fs->sector_size;
-		_nvs_entry_sector_advance(fs);
-		rc = _nvs_gc(fs);
-		if (rc) {
-			goto err;
-		}
+	/* Difference in length or illegal length- bad entry, copy part that
+	 * is  OK
+	 */
+	len1 = fs->write_location & (fs->sector_size - 1);
+	addr = fs->write_location & ~(fs->sector_size - 1);
+	/* Go to next sector and erase it to be sure */
+	_nvs_addr_advance(fs, &addr, fs->sector_size);
+	/* erase the sector */
+	rc = _nvs_flash_erase(fs, addr, fs->sector_size);
+	if (rc) {
+		goto err;
+	}
+	fs->write_location = addr;
+	_nvs_addr_decrease(fs, &addr, fs->sector_size);
+	/* copy the OK data to temporary sector */
+	_nvs_flash_block_move(fs, addr, len1);
```

Comment:
Is it up to the application that the touched data are still valid?

Reply:
Just have idea that we can return some value which signals that recovery was done and especial the last entry is not trusted. We can expected the NOR flash (so the SoC embedded flash) is reliable regards keeping its content - so it dose make sens to verify the content rare.